src/LuaSerializer.cpp

```diff
@@ -328,6 +328,261 @@ const char *LuaSerializer::unpickle(lua_State *l, const char *pos)
+
+			lua_getfield(l, LUA_REGISTRYINDEX, "PiSerializerClasses");
+
+			lua_getfield(l, -1, cl);
+			if (lua_isnil(l, -1))
+				luaL_error(l, "No Serialize method found for class '%s'\n", cl);
+
+			lua_getfield(l, -1, "Serialize");
+			if (lua_isnil(l, -1))
+				luaL_error(l, "No Serialize method found for class '%s'\n", cl);
+
+			lua_pushvalue(l, idx);
+			pi_lua_protected_call(l, 1, 1);
+
+			idx = lua_gettop(l);
+
+			if (lua_isnil(l, idx)) {
+				lua_pop(l, 5);
+				LUA_DEBUG_END(l, 0);
+				return;
+			}
+		}
+	}
+
+	switch (lua_type(l, idx)) {
+		case LUA_TNIL:
+			break;
+
+		case LUA_TNUMBER: {
+			out = Json::Value(lua_tonumber(l, idx));
+			break;
+		}
+
+		case LUA_TBOOLEAN: {
+			out = Json::Value(static_cast<bool>(lua_toboolean(l, idx)));
+			break;
+		}
+
+		case LUA_TSTRING: {
+			lua_pushvalue(l, idx);
+			size_t len;
+			const char *str = lua_tolstring(l, -1, &len);
+			out = Json::Value(str, str + len);
+			lua_pop(l, 1);
+			break;
+		}
+
+		case LUA_TTABLE: {
+			lua_Integer ptr = lua_Integer(lua_topointer(l, to_serialize));
+			lua_pushinteger(l, ptr);                                        // ptr
+
+			lua_getfield(l, LUA_REGISTRYINDEX, "PiSerializerTableRefs");    // ptr reftable
+			lua_pushvalue(l, -2);                                           // ptr reftable ptr
+			lua_rawget(l, -2);                                              // ptr reftable ???
+
+			out["inner_ref"] = Json::Value(Json::Int64(ptr));
+
+			if (!lua_isnil(l, -1)) {
+				lua_pop(l, 3);                                              // [empty]
+			} else {
+				lua_pushvalue(l, -3);                                       // ptr reftable nil ptr
+				lua_pushvalue(l, to_serialize);                             // ptr reftable nil ptr table
+				lua_rawset(l, -4);                                          // ptr reftable nil
+				lua_pop(l, 3);                                              // [empty]
+
+				Json::Value inner(Json::arrayValue);
+
+				lua_pushvalue(l, idx);
+				lua_pushnil(l);
+				while (lua_next(l, -2)) {
+					lua_pushvalue(l, -2);
+					const char *k = lua_tostring(l, -1);
+					std::string new_key = key + "." + (k? std::string(k) : "<" + std::string(lua_typename(l, lua_type(l, -1))) + ">");
+					lua_pop(l, 1);
+
+					// Copy the values to pickle, as they might be mutated by the pickling process.
+					Json::Value out_k, out_v;
+
+					pickle_json(l, -2, out_k, new_key);
+					pickle_json(l, -1, out_v, new_key);
+
+					inner.append(out_k);
+					inner.append(out_v);
+
+					lua_pop(l, 1);
+				}
+				lua_pop(l, 1);
+
+				out["inner"] = Json::Value(inner);
+			}
+
+			break;
+		}
+
+		case LUA_TUSERDATA: {
+			LuaObjectBase *lo = static_cast<LuaObjectBase*>(lua_touserdata(l, idx));
+			void *o = lo->GetObject();
+			if (!o)
+				Error("Lua serializer '%s' tried to serialize an invalid '%s' object", key.c_str(), lo->GetType());
+
+			const std::string pickled = lo->Serialize();
+			BinStrToJson(out, pickled, "udata");
```

Comment:
Might need a LuaObject::SerializeToJSON method (and corresponding DeserializeFromJSON).